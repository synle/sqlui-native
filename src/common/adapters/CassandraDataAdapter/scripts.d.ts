import BaseDataScript from 'src/common/adapters/BaseDataAdapter/scripts';
import { SqlAction } from 'typings';
export declare function getSelectAllColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getSelectSpecificColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getInsert(input: SqlAction.TableInput, value?: Record<string, any>): SqlAction.Output | undefined;
export declare function getBulkInsert(input: SqlAction.TableInput, rows?: Record<string, any>[]): SqlAction.Output | undefined;
export declare function getUpdateWithValues(input: SqlAction.TableInput, value: Record<string, any>, conditions: Record<string, any>): SqlAction.Output | undefined;
export declare function getUpdate(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDelete(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getCreateKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getDropKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getCreateConnectionDatabase(input: SqlAction.ConnectionInput): SqlAction.Output | undefined;
export declare function getCreateTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getAddColumn(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare class ConcreteDataScripts extends BaseDataScript {
    dialects: string[];
    getIsTableIdRequiredForQuery(): boolean;
    getSyntaxMode(): string;
    supportMigration(): boolean;
    supportCreateRecordForm(): boolean;
    supportEditRecordForm(): boolean;
    getTableScripts(): (typeof getInsert)[];
    getDatabaseScripts(): (typeof getCreateKeyspace)[];
    getConnectionScripts(): (typeof getCreateConnectionDatabase)[];
    getSampleConnectionString(dialect: any): string;
    getSampleSelectQuery(tableActionInput: any): SqlAction.Output;
    getCodeSnippet(connection: any, query: any, language: any): string;
}
declare const _default: ConcreteDataScripts;
export default _default;
