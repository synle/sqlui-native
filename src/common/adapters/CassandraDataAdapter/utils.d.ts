declare type ClientOptions = {
    host: string;
    port: number;
    contactPoints: string[];
    keyspace?: string;
    authProvider?: {
        username: string;
        password: string;
    };
};
export declare function getClientOptions(connectionOption: string, database?: string): ClientOptions;
export {};
