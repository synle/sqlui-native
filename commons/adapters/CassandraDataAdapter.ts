import * as cassandra from 'cassandra-driver';
import { SqluiCore } from '../../typings';
import IDataAdapter from './IDataAdapter';
import BaseDataAdapter from './BaseDataAdapter';

export default class CassandraAdapter extends BaseDataAdapter implements IDataAdapter {
  dialect: SqluiCore.Dialect = 'cassandra';
  /**
   * cassandra version
   * @type {number}
   */
  version?: string;
  isCassandra2?: boolean;

  constructor(connectionOption: string) {
    super(connectionOption);
    this.dialect = 'cassandra';
  }

  private async getConnection(database?: string): Promise<cassandra.Client> {
    // attempt to pull in connections
    return new Promise<cassandra.Client>(async (resolve, reject) => {
      try {
        const connectionParameters = BaseDataAdapter.getConnectionParameters(this.connectionOption);

        const connectionHosts = connectionParameters?.hosts || [];
        if(connectionHosts.length === 0){
          // we need a host in the connection string
          reject('Invalid connection. Host and Port not found');
        }

        //@ts-ignore
        const client = new cassandra.Client({
          contactPoints: [connectionHosts[0].host],
          protocolOptions: {
            port: connectionHosts[0].port,
          },
          keyspace: database,
        });

        await this.authenticate(client);

        resolve(client);
      } catch (err) {
        reject(err);
      }
    });
  }

  async authenticate(client?: cassandra.Client) {
    // TODO: To Be Implemented
    return new Promise<void>((resolve, reject) => {
      client?.connect((err: unknown) => {
        if (err) {
          client?.shutdown();
          return reject(err);
        }

        this.version = client?.getState().getConnectedHosts()[0].cassandraVersion;
        this.isCassandra2 = this.version === '2';

        resolve();
      });
    });
  }

  async getDatabases(): Promise<SqluiCore.DatabaseMetaData[]> {
    const client = await this.getConnection();

    //@ts-ignore
    const keyspaces = Object.keys(client?.metadata?.keyspaces);
    return keyspaces
      .map((keyspace) => ({
        name: keyspace,
        tables: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getTables(database?: string): Promise<SqluiCore.TableMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
          SELECT columnfamily_name as name
          FROM system.schema_columnfamilies
          WHERE keyspace_name = ?
        `;
    } else {
      sql = `
          SELECT table_name as name
          FROM system_schema.tables
          WHERE keyspace_name = ?
        `;
    }

    const res = await this._execute(sql, [database]);

    return res.rows
      .map((row) => ({
        name: row.name,
        columns: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getColumns(table: string, database?: string): Promise<SqluiCore.ColumnMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
        SELECT type as position, column_name, validator as type
        FROM system.schema_columns
        WHERE keyspace_name = ?
          AND columnfamily_name = ?
      `;
    } else {
      sql = `
        SELECT position, column_name, type
        FROM system_schema.columns
        WHERE keyspace_name = ?
          AND table_name = ?
      `;
    }
    const res = await this._execute(sql, [database, table]);

    return res.rows
      .map((row) => ({
        name: row.column_name,
        type: row.type,
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  private async _execute(sql: string, params?: string[], database?: string) {
    const client = await this.getConnection(database);
    try {
      let res = await client.execute(sql, params || [], { prepare: true });
      client?.shutdown();
      return res;
    } catch (err) {
      client?.shutdown();
      throw err;
    }
  }

  async execute(sql: string, database?: string): Promise<SqluiCore.Result> {
    let rawToUse: any | undefined;
    let metaToUse: any | undefined;

    try {
      const res = await this._execute(sql, undefined, database);

      if (!res) {
        return {
          ok: false,
          raw: [],
          meta: null,
        };
      }
      const { rows } = res;

      //@ts-ignore
      delete res.rows;

      rawToUse = rows;
      metaToUse = res;

      return {
        ok: true,
        raw: rawToUse,
        meta: metaToUse,
      };
    } catch (error) {
      return {
        ok: false,
        error,
      };
    }
  }
}
