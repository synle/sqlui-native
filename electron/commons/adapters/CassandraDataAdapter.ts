import * as cassandra from 'cassandra-driver';
import { SqluiCore } from 'typings';
import IDataAdapter from 'electron/commons/adapters/IDataAdapter';
import BaseDataAdapter from 'electron/commons/adapters/BaseDataAdapter';

export default class CassandraDataAdapter extends BaseDataAdapter implements IDataAdapter {
  dialect: SqluiCore.Dialect = 'cassandra';
  /**
   * cassandra version
   * @type {number}
   */
  version?: string;
  isCassandra2?: boolean;

  constructor(connectionOption: string) {
    super(connectionOption);
  }

  private async getConnection(database?: string): Promise<cassandra.Client> {
    // attempt to pull in connections
    return new Promise<cassandra.Client>(async (resolve, reject) => {
      try {
        const connectionParameters = BaseDataAdapter.getConnectionParameters(this.connectionOption);

        const connectionHosts = connectionParameters?.hosts || [];
        if (connectionHosts.length === 0) {
          // we need a host in the connection string
          reject('Invalid connection. Host and Port not found');
        }

        const clientOptions: cassandra.ClientOptions = {
          contactPoints: [connectionHosts[0].host],
          protocolOptions: {
            port: connectionHosts[0].port,
          },
          keyspace: database,
        };

        // client authentication
        let authProvider: cassandra.auth.PlainTextAuthProvider | undefined;
        if (connectionParameters?.username && connectionParameters?.password) {
          clientOptions.authProvider = new cassandra.auth.PlainTextAuthProvider(
            connectionParameters?.username,
            connectionParameters?.password,
          );
        }

        const client = new cassandra.Client(clientOptions);
        await this.authenticateClient(client);

        resolve(client);
      } catch (err) {
        reject(err);
      }
    });
  }

  private async authenticateClient(client?: cassandra.Client) {
    // TODO: To Be Implemented
    return new Promise<void>((resolve, reject) => {
      client?.connect((err: unknown) => {
        if (err) {
          client?.shutdown();
          return reject(err);
        }

        this.version = client?.getState().getConnectedHosts()[0].cassandraVersion;
        this.isCassandra2 = this.version.indexOf('2.') === 0;

        resolve();
      });
    });
  }

  async authenticate() {
    await this.getConnection();
  }

  async getDatabases(): Promise<SqluiCore.DatabaseMetaData[]> {
    const client = await this.getConnection();

    //@ts-ignore
    const keyspaces = Object.keys(client?.metadata?.keyspaces);
    return keyspaces
      .map((keyspace) => ({
        name: keyspace,
        tables: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getTables(database?: string): Promise<SqluiCore.TableMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
          SELECT columnfamily_name as name
          FROM system.schema_columnfamilies
          WHERE keyspace_name = ?
        `;
    } else {
      sql = `
          SELECT table_name as name
          FROM system_schema.tables
          WHERE keyspace_name = ?
        `;
    }

    const res = await this._execute(sql, [database]);

    return res.rows
      .map((row) => ({
        name: row.name,
        columns: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getColumns(table: string, database?: string): Promise<SqluiCore.ColumnMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
        SELECT type as position, column_name as name, validator as type, type as kind
        FROM system.schema_columns
        WHERE keyspace_name = ?
          AND columnfamily_name = ?
      `;
    } else {
      sql = `
        SELECT position, column_name as name, type, kind
        FROM system_schema.columns
        WHERE keyspace_name = ?
          AND table_name = ?
      `;
    }
    const res = await this._execute(sql, [database, table]);

    return res.rows
      .sort((a: any, b: any) => (a.name || '').localeCompare(b.name || ''))
      .map((row) => ({
        name: row.name,
        type: row.type,
        kind: row.kind,
      }));
  }

  private async _execute(sql: string, params?: string[], database?: string) {
    const client = await this.getConnection(database);
    try {
      let res = await client.execute(sql, params || [], { prepare: true });
      client?.shutdown();
      return res;
    } catch (err) {
      client?.shutdown();
      throw err;
    }
  }

  async execute(sql: string, database?: string): Promise<SqluiCore.Result> {
    let rawToUse: any | undefined;
    let metaToUse: any | undefined;

    try {
      const res = await this._execute(sql, undefined, database);

      if (!res) {
        return {
          ok: false,
          raw: [],
          meta: null,
        };
      }
      const { rows } = res;

      //@ts-ignore
      delete res.rows;

      rawToUse = rows;
      metaToUse = res;

      return {
        ok: true,
        raw: rawToUse,
        meta: metaToUse,
      };
    } catch (error) {
      return {
        ok: false,
        error,
      };
    }
  }
}
