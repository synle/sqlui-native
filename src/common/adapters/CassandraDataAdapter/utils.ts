import BaseDataAdapter from 'src/common/adapters/BaseDataAdapter/index';

type ClientOptions = {
  host: string;
  port: number;
  contactPoints: string[];
  keyspace?: string;
  authProvider?: {
    username: string;
    password: string;
  };
};

export function getClientOptions(connectionOption: string, database?: string): ClientOptions {
  const connectionParameters = BaseDataAdapter.getConnectionParameters(connectionOption);

  const connectionHosts = connectionParameters?.hosts || [];
  if (connectionHosts.length === 0) {
    // we need a host in the connection string
    throw 'Invalid connection. Host and Port not found';
  }

  const host = connectionHosts[0].host;
  const port = connectionHosts[0].port || 9042;

  const clientOptions: ClientOptions = {
    host,
    port,
    contactPoints: [`${host}:${port}`],
  };

  if (database) {
    clientOptions.keyspace = database;
  }

  // client authentication
  if (connectionParameters?.username && connectionParameters?.password) {
    clientOptions.authProvider = {
      username: connectionParameters?.username,
      password: connectionParameters?.password,
    };
  }

  return clientOptions;
}