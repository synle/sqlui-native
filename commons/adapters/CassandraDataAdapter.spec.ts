import CassandraDataAdapter from './CassandraDataAdapter';

const adapter = new CassandraDataAdapter('cassandra://test.db');

describe('cassandra', () => {
  test('cassandra - Get database', async () => {
    const databases = await adapter.getDatabases();
    expect(databases).toMatchSnapshot();
  });

  test('cassandra - Get tables', async () => {
    const tables = await adapter.getTables('system');
    expect(tables).toMatchSnapshot();
  });

  test('cassandra - Get columns', async () => {
    const columns = await adapter.getColumns('columns', 'system_schema');
    expect(columns).toMatchSnapshot();
  });

  test('Execute Select', async () => {
    const resp = await adapter.execute(`SELECT * FROM tables LIMIT 10`, 'system_schema');
    expect(resp).toMatchSnapshot();
  });
});
