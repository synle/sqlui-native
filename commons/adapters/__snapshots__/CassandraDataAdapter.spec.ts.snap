// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cassandra Execute Select 1`] = `
Object {
  "meta": ResultSet {
    "columns": Array [
      Object {
        "name": "keyspace_name",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
      Object {
        "name": "table_name",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
      Object {
        "name": "additional_write_policy",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
      Object {
        "name": "bloom_filter_fp_chance",
        "type": Object {
          "code": 7,
          "type": null,
        },
      },
      Object {
        "name": "caching",
        "type": Object {
          "code": 33,
          "info": Array [
            Object {
              "code": 13,
              "type": null,
            },
            Object {
              "code": 13,
              "type": null,
            },
          ],
          "type": null,
        },
      },
      Object {
        "name": "cdc",
        "type": Object {
          "code": 4,
          "type": null,
        },
      },
      Object {
        "name": "comment",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
      Object {
        "name": "compaction",
        "type": Object {
          "code": 33,
          "info": Array [
            Object {
              "code": 13,
              "type": null,
            },
            Object {
              "code": 13,
              "type": null,
            },
          ],
          "type": null,
        },
      },
      Object {
        "name": "compression",
        "type": Object {
          "code": 33,
          "info": Array [
            Object {
              "code": 13,
              "type": null,
            },
            Object {
              "code": 13,
              "type": null,
            },
          ],
          "type": null,
        },
      },
      Object {
        "name": "crc_check_chance",
        "type": Object {
          "code": 7,
          "type": null,
        },
      },
      Object {
        "name": "dclocal_read_repair_chance",
        "type": Object {
          "code": 7,
          "type": null,
        },
      },
      Object {
        "name": "default_time_to_live",
        "type": Object {
          "code": 9,
          "type": null,
        },
      },
      Object {
        "name": "extensions",
        "type": Object {
          "code": 33,
          "info": Array [
            Object {
              "code": 13,
              "type": null,
            },
            Object {
              "code": 3,
              "type": null,
            },
          ],
          "type": null,
        },
      },
      Object {
        "name": "flags",
        "type": Object {
          "code": 34,
          "info": Object {
            "code": 13,
            "type": null,
          },
          "type": null,
        },
      },
      Object {
        "name": "gc_grace_seconds",
        "type": Object {
          "code": 9,
          "type": null,
        },
      },
      Object {
        "name": "id",
        "type": Object {
          "code": 12,
          "type": null,
        },
      },
      Object {
        "name": "max_index_interval",
        "type": Object {
          "code": 9,
          "type": null,
        },
      },
      Object {
        "name": "memtable_flush_period_in_ms",
        "type": Object {
          "code": 9,
          "type": null,
        },
      },
      Object {
        "name": "min_index_interval",
        "type": Object {
          "code": 9,
          "type": null,
        },
      },
      Object {
        "name": "read_repair",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
      Object {
        "name": "read_repair_chance",
        "type": Object {
          "code": 7,
          "type": null,
        },
      },
      Object {
        "name": "speculative_retry",
        "type": Object {
          "code": 13,
          "type": null,
        },
      },
    ],
    "info": Object {
      "achievedConsistency": 10,
      "customPayload": undefined,
      "queriedHost": "127.0.0.1:9042",
      "speculativeExecutions": 0,
      "traceId": undefined,
      "triedHosts": Object {
        "127.0.0.1:9042": null,
      },
      "warnings": undefined,
    },
    "nextPage": undefined,
    "pageState": null,
    "rowLength": 10,
  },
  "ok": false,
  "raw": Array [
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "user network permissions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 7776000,
      "id": "d46780c2-2f1c-3db9-b4c1-b8d9fbc0cc23",
      "keyspace_name": "system_auth",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 0,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "network_permissions",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "index of db roles with permissions granted on a resource",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 7776000,
      "id": "5f2fbdad-91f1-3946-bd25-d5da3a5c35ec",
      "keyspace_name": "system_auth",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 0,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "resource_role_permissons_index",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "role memberships lookup table",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 7776000,
      "id": "0ecdaa87-f8fb-3e60-88d1-74fb36fe5c0d",
      "keyspace_name": "system_auth",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 0,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "role_members",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "permissions granted to db roles",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 7776000,
      "id": "3afbe79f-2194-31a7-add7-f5ab90d8ec9c",
      "keyspace_name": "system_auth",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 0,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "role_permissions",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "role definitions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 7776000,
      "id": "5bc52802-de25-35ed-aeab-188eecebb090",
      "keyspace_name": "system_auth",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 0,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "roles",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "user defined aggregate definitions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 604800,
      "id": "924c5587-2e3a-345b-b10c-12f37c1ba895",
      "keyspace_name": "system_schema",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 3600000,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "aggregates",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "column definitions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 604800,
      "id": "24101c25-a2ae-3af7-87c1-b40ee1aca33f",
      "keyspace_name": "system_schema",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 3600000,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "columns",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "dropped column registry",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 604800,
      "id": "5e7583b5-f3f4-3af1-9a39-b7e1d6f5f11f",
      "keyspace_name": "system_schema",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 3600000,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "dropped_columns",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "user defined function definitions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 604800,
      "id": "96489b79-80be-3e14-a701-66a0b9159450",
      "keyspace_name": "system_schema",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 3600000,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "functions",
    },
    Row {
      "additional_write_policy": "99p",
      "bloom_filter_fp_chance": 0.01,
      "caching": Object {
        "keys": "ALL",
        "rows_per_partition": "NONE",
      },
      "cdc": null,
      "comment": "secondary index definitions",
      "compaction": Object {
        "class": "org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy",
        "max_threshold": "32",
        "min_threshold": "4",
      },
      "compression": Object {
        "chunk_length_in_kb": "16",
        "class": "org.apache.cassandra.io.compress.LZ4Compressor",
      },
      "crc_check_chance": 1,
      "dclocal_read_repair_chance": 0,
      "default_time_to_live": 0,
      "extensions": Object {},
      "flags": Array [
        "compound",
      ],
      "gc_grace_seconds": 604800,
      "id": "0feb57ac-311f-382f-ba6d-9024d305702f",
      "keyspace_name": "system_schema",
      "max_index_interval": 2048,
      "memtable_flush_period_in_ms": 3600000,
      "min_index_interval": 128,
      "read_repair": "BLOCKING",
      "read_repair_chance": 0,
      "speculative_retry": "99p",
      "table_name": "indexes",
    },
  ],
}
`;

exports[`cassandra cassandra - Get columns 1`] = `
Array [
  Object {
    "name": "clustering_order",
    "type": "text",
  },
  Object {
    "name": "column_name",
    "type": "text",
  },
  Object {
    "name": "column_name_bytes",
    "type": "blob",
  },
  Object {
    "name": "keyspace_name",
    "type": "text",
  },
  Object {
    "name": "kind",
    "type": "text",
  },
  Object {
    "name": "position",
    "type": "int",
  },
  Object {
    "name": "table_name",
    "type": "text",
  },
  Object {
    "name": "type",
    "type": "text",
  },
]
`;

exports[`cassandra cassandra - Get database 1`] = `
Array [
  Object {
    "name": "system",
    "tables": Array [],
  },
  Object {
    "name": "system_auth",
    "tables": Array [],
  },
  Object {
    "name": "system_distributed",
    "tables": Array [],
  },
  Object {
    "name": "system_schema",
    "tables": Array [],
  },
  Object {
    "name": "system_traces",
    "tables": Array [],
  },
  Object {
    "name": "system_views",
    "tables": Array [],
  },
  Object {
    "name": "system_virtual_schema",
    "tables": Array [],
  },
]
`;

exports[`cassandra cassandra - Get tables 1`] = `
Array [
  Object {
    "columns": Array [],
    "name": "available_ranges",
  },
  Object {
    "columns": Array [],
    "name": "available_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "batches",
  },
  Object {
    "columns": Array [],
    "name": "built_views",
  },
  Object {
    "columns": Array [],
    "name": "compaction_history",
  },
  Object {
    "columns": Array [],
    "name": "IndexInfo",
  },
  Object {
    "columns": Array [],
    "name": "local",
  },
  Object {
    "columns": Array [],
    "name": "paxos",
  },
  Object {
    "columns": Array [],
    "name": "peer_events",
  },
  Object {
    "columns": Array [],
    "name": "peer_events_v2",
  },
  Object {
    "columns": Array [],
    "name": "peers",
  },
  Object {
    "columns": Array [],
    "name": "peers_v2",
  },
  Object {
    "columns": Array [],
    "name": "prepared_statements",
  },
  Object {
    "columns": Array [],
    "name": "repairs",
  },
  Object {
    "columns": Array [],
    "name": "size_estimates",
  },
  Object {
    "columns": Array [],
    "name": "sstable_activity",
  },
  Object {
    "columns": Array [],
    "name": "table_estimates",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "view_builds_in_progress",
  },
]
`;
